tests/prefetch_related/tests.py

```diff
@@ -2073,3 +2074,197 @@ def test_empty_order(self):
+        self.assertSetEqual(
+            set(gp.favorite_authors.all()),
+            {self.related1, self.related2, self.related3, self.related4},
+        )
+        self.assertSetEqual(
+            set(gp.authorwithage.favorite_authors.all()),
+            {self.related1, self.related2, self.related3, self.related4},
+        )
+
+    def test_add_clears_prefetched_objects_in_grandparent(self):
+        gp = (
+            Author.objects.select_related(
+                "authorwithage", "authorwithage__authorwithagechild"
+            )
+            .prefetch_related("favorite_authors")
+            .get(pk=self.m2m_child.pk)
+        )
+        self.assertSetEqual(
+            set(gp.favorite_authors.all()),
+            {self.related1, self.related2, self.related3},
+        )
+        self.assertSetEqual(
+            set(gp.authorwithage.favorite_authors.all()),
+            {self.related1, self.related2, self.related3},
+        )
+        self.assertSetEqual(
+            set(gp.authorwithage.authorwithagechild.favorite_authors.all()),
+            {self.related1, self.related2, self.related3},
+        )
+        gp.authorwithage.authorwithagechild.favorite_authors.add(self.related4)
+        self.assertSetEqual(
+            set(gp.favorite_authors.all()),
+            {self.related1, self.related2, self.related3, self.related4},
+        )
+        self.assertSetEqual(
+            set(gp.authorwithage.favorite_authors.all()),
+            {self.related1, self.related2, self.related3, self.related4},
+        )
+        self.assertSetEqual(
+            set(gp.authorwithage.authorwithagechild.favorite_authors.all()),
+            {self.related1, self.related2, self.related3, self.related4},
+        )
+
+    def test_remove_clears_prefetched_objects_in_parent(self):
+        gp = (
+            Author.objects.select_related("authorwithage")
+            .prefetch_related("favorite_authors")
+            .get(pk=self.m2m_child.pk)
+        )
+        self.assertSetEqual(
+            set(gp.favorite_authors.all()),
+            {self.related1, self.related2, self.related3},
+        )
+        self.assertSetEqual(
+            set(gp.authorwithage.favorite_authors.all()),
+            {self.related1, self.related2, self.related3},
+        )
+        gp.authorwithage.favorite_authors.clear()
+        self.assertListEqual(list(gp.favorite_authors.all()), [])
+        self.assertListEqual(list(gp.authorwithage.favorite_authors.all()), [])
+
+    def test_remove_clears_prefetched_objects_in_grandparent(self):
+        gp = (
+            Author.objects.select_related(
+                "authorwithage", "authorwithage__authorwithagechild"
+            )
+            .prefetch_related("favorite_authors")
+            .get(pk=self.m2m_child.pk)
+        )
+        self.assertSetEqual(
+            set(gp.favorite_authors.all()),
+            {self.related1, self.related2, self.related3},
+        )
```

Comment:
We usually stick to `assertQuerySetEqual` for these cases

```suggestion
        self.assertSetEqual(
            gp.favorite_authors.all()
            [self.related1, self.related2, self.related3],
        )
```

Reply:
Fixed!